Building with lld and download-ci-llvm

I order to try local patches on a cross-compiled project, I built with `./x.py build --target x86_64-unknown-linux-gnu,riscv32i-unknown-none-elf` then, in my project set `rust-toolchain.toml` to `path = "/home/simon/projects/rust/build/x86_64-unknown-linux-gnu/stage1/"`. The first error message was:

```
  = note: Could not find tool: lld
          at: /home/simon/projects/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin/rust-lld
          Consider `rustup component add llvm-tools-preview`
```

It was not obvious in `./x.py build --help` or `config.toml.example` how to build `llvm-tools-preview` for a local toolchain, but folks on Zulip pointed me to adding `[rust] lld = true` to `config.toml`.

Rebuilding the toolchain failed with:

```
Writer.cpp:(.text._ZN3lld4wasm12_GLOBAL__N_16Writer3runEv+0x23f8): undefined reference to `llvm::parallelForEachN(unsigned long, unsigned long, llvm::function_ref<void (unsigned long)>)'
```

My `config.toml` also had `profile = "compiler"` which enables `download-ci-llvm = "if-available"`. It looks like building LLD from source doesn’t work when LLVM artifacts are downloaded from CI.

This configuration should either download LLD from CI too, or print a better error message.

--- a/src/bootstrap/dist.rs
+++ b/src/bootstrap/dist.rs
@@ -2067,6 +2067,12 @@ impl Step for RustDev {
 
         builder.ensure(crate::native::Llvm { target });
 
+        // If the config has LLD enabled, ensure its step. We'll we want to package it, and use it
+        // in download-ci-llvm.
+        if builder.config.lld_enabled {
+            builder.ensure(crate::native::Lld { target });
+        }
+
         let src_bindir = builder.llvm_out(target).join("bin");
         // If updating this list, you likely want to change
         // src/bootstrap/download-ci-llvm-stamp as well, otherwise local users

--- a/src/bootstrap/dist.rs
+++ b/src/bootstrap/dist.rs
@@ -2067,11 +2067,8 @@ impl Step for RustDev {
 
         builder.ensure(crate::native::Llvm { target });
 
-        // If the config has LLD enabled, ensure its step. We'll we want to package it, and use it
-        // in download-ci-llvm.
-        if builder.config.lld_enabled {
-            builder.ensure(crate::native::Lld { target });
-        }
+        // We want to package `lld` to use it with `download-ci-llvm`.
+        builder.ensure(crate::native::Lld { target });
 
         let src_bindir = builder.llvm_out(target).join("bin");
         // If updating this list, you likely want to change

--- a/src/bootstrap/download-ci-llvm-stamp
+++ b/src/bootstrap/download-ci-llvm-stamp
@@ -1,4 +1,4 @@
 Change this file to make users of the `download-ci-llvm` configuration download
 a new version of LLVM from CI, even if the LLVM submodule hasn’t changed.
 
-Last change is for: https://github.com/rust-lang/rust/pull/102790
+Last change is for: https://github.com/rust-lang/rust/pull/104748

--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -823,8 +823,7 @@ impl Step for Lld {
         }
         let target = self.target;
 
-        let LlvmResult { llvm_config, llvm_cmake_dir } =
-            builder.ensure(Llvm { target: self.target });
+        let LlvmResult { llvm_config, llvm_cmake_dir } = builder.ensure(Llvm { target });
 
         // The `dist` step packages LLD next to LLVM's binaries for download-ci-llvm. The root path
         // we usually expect here is `./build/$triple/ci-llvm/`, with the binaries in its `bin`

--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -826,6 +826,20 @@ impl Step for Lld {
         let LlvmResult { llvm_config, llvm_cmake_dir } =
             builder.ensure(Llvm { target: self.target });
 
+        // The `dist` step packages LLD next to LLVM's binaries for download-ci-llvm. The root path
+        // we usually expect here is `./build/$triple/ci-llvm/`, with the binaries in its `bin`
+        // subfolder. We check if that's the case, and if LLD's binary already exists there next to
+        // `llvm-config`: if so, we can use it instead of building LLVM/LLD from source.
+        let ci_llvm_bin = llvm_config.parent().unwrap();
+        if ci_llvm_bin.is_dir() && ci_llvm_bin.file_name().unwrap() == "bin" {
+            let lld_path = ci_llvm_bin.join(exe("lld", target));
+            if lld_path.exists() {
+                // The following steps copying `lld` as `rust-lld` to the sysroot, expect it in the
+                // `bin` subfolder of this step's out dir.
+                return ci_llvm_bin.parent().unwrap().to_path_buf();
+            }
+        }
+
         let out_dir = builder.lld_out(target);
         let done_stamp = out_dir.join("lld-finished-building");
         if done_stamp.exists() {

--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -63,13 +63,13 @@ impl LdFlags {
     }
 }
 
-// This returns whether we've already previously built LLVM.
-//
-// It's used to avoid busting caches during x.py check -- if we've already built
-// LLVM, it's fine for us to not try to avoid doing so.
-//
-// This will return the llvm-config if it can get it (but it will not build it
-// if not).
+/// This returns whether we've already previously built LLVM.
+///
+/// It's used to avoid busting caches during x.py check -- if we've already built
+/// LLVM, it's fine for us to not try to avoid doing so.
+///
+/// This will return the llvm-config if it can get it (but it will not build it
+/// if not).
 pub fn prebuilt_llvm_config(
     builder: &Builder<'_>,
     target: TargetSelection,

